### PR TITLE
Rewrote "How Safe and Unsafe Interact" Nomicon chapter.

### PR DESCRIPTION
**Pull request by alilleybrinker**
The previous version of the chapter covered a lot of ground, but was a little meandering and hard to follow at times. This draft is intended to be clearer and more direct, while still providing the same information as the previous version.

r? @steveklabnik

### CODE REVIEW

**Comment by AndrewBrinker** [comment]
Here's the [current version](https://doc.rust-lang.org/nomicon/safe-unsafe-meaning.html) of the rewritten chapter, for convenience.

File: src/doc/nomicon/safe-unsafe-meaning.md [hunk 1]
```diff
@@ -1,150 +1,131 @@
...
+between types which can "just" be compared, and those that provide a total
+ordering (where every value of the type is either equal to, greater than,
+or less than any other value of the same type). The sorted map type
+`BTreeMap` doesn't make sense for partially-ordered types, and so it
+requires that any key type for it implements the `Ord` trait. However,
+`BTreeMap` has Unsafe Rust code inside of its implementation, and this
+Unsafe Rust code cannot assume that any `Ord` implementation it gets makes
+sense. The unsafe portions of `BTreeMap`'s internals have to be careful to
+maintain all necessary contracts, even if a key type's `Ord` implementation
+is completely nonsensical.
```
Comment (pnkfelix) [inline src/doc/nomicon/safe-unsafe-meaning.md]: Be more specific here, and say "even if a key type's Ord implementation does not implement a total ordering"

**Comment by tshepang** [comment]
please squash

**Comment by AndrewBrinker** [comment]
Alright, I've squashed.

**Comment by steveklabnik** [comment]
Hey, sorry for taking a while to review this. Will try to get to it tomorrow.

**Comment by AndrewBrinker** [comment]
No problem!

**Comment by steveklabnik** [comment]
@AndrewBrinker so, since this is a pretty significant re-write, and it's hard to tell what's different and what's changed, I'm going to review this from the position of a brand-new chapter. Some of which, you inherited, of course :)

If you don't have the time/desire to make these kinds of tweaks, let me know, and we don't have to block the PR on them, but I might need your help in separating the past from the present, given the diff is so unhelpful here. :smile:

File: src/doc/nomicon/safe-unsafe-meaning.md [hunk 2]
```diff
@@ -1,150 +1,131 @@
...
-* `mem::transmute` reinterprets some value as having the given type,
-  bypassing type safety in arbitrary ways. (see [conversions] for details)
-* All FFI functions are `unsafe` because they can do arbitrary things.
-  C being an obvious culprit, but generally any language can do something
-  that Rust isn't happy about.
+What's the relationship between Safe Rust and Unsafe Rust? How do they
+interact?
+
+The separation between Safe Rust and Unsafe Rust is controlled with the
+`unsafe` keyword, which acts as a sort of *foreign function interface*
```
Comment (steveklabnik) [inline src/doc/nomicon/safe-unsafe-meaning.md]: I am a little uncomfortable with overloading this term in this way. It's not the worst, but I wonder if we can not make this implication so strong. Then again, maybe it is apt; `unsafe` is a part of the functions' type, so in some sense, they are different... i dunno.

The reason I'm a bit hesitant is this invokes something like "wait, is the calling convention different?"

File: src/doc/nomicon/safe-unsafe-meaning.md [hunk 3]
```diff
@@ -1,150 +1,131 @@
...
+from one to the other. This boundary is why we can say Safe Rust is a
+safe language: all the unsafe parts are kept exclusively behind the FFI
+boundary, *just like any other safe language*. Best of all, because Safe
+Rust is a subset of Unsafe Rust, the two can be cleanly intermixed,
+without headers, runtimes, or any other FFI boilerplate.
```
Comment (steveklabnik) [inline src/doc/nomicon/safe-unsafe-meaning.md]: This does sort of back off a bit, in that way. So maybe this is okay.

File: src/doc/nomicon/safe-unsafe-meaning.md [hunk 4]
```diff
@@ -1,150 +1,131 @@
...
+
+The `unsafe` keyword has dual purposes: to declare the existence of
```
Comment (steveklabnik) [inline src/doc/nomicon/safe-unsafe-meaning.md]: Extreme nitpick: this should be 

> The `unsafe` keyword has two uses:

or

> The `unsafe` keyword is dual-purpose:

File: src/doc/nomicon/safe-unsafe-meaning.md [hunk 5]
```diff
@@ -1,150 +1,131 @@
...
+contracts the compiler can't check, and to declare that the adherence
+of some code to those contracts has been checked by the programmer,
+and the code can therefore be trusted.
```
Comment (steveklabnik) [inline src/doc/nomicon/safe-unsafe-meaning.md]: I would leave this last clause off. When reading this, it feels like a third purpose, even if it's not.

File: src/doc/nomicon/safe-unsafe-meaning.md (hunk 2 above)
Comment (AndrewBrinker) [inline src/doc/nomicon/safe-unsafe-meaning.md]: Yeah, it's a bit of an overloaded term in this context. I am happy to take a crack at a different wording tonight.

File: src/doc/nomicon/safe-unsafe-meaning.md (hunk 5 above)
Comment (AndrewBrinker) [inline src/doc/nomicon/safe-unsafe-meaning.md]: Yeah, good call.

**Comment by AndrewBrinker** [comment]
I've rewritten the second paragraph to eliminate the FFI bits, and I think it's actually clearer now. I also made the smaller changes you suggested.

As far as what's different between this and the original, I'll start by noting that my goal with this was basically to reword and simplify the original, so there are a lot of similarities.

The first paragraph ("What's the relationship...") is the same in both versions.

The second paragraph in the new version ("The separation between...") is a combination of paragraphs 2 and 3 from the original, with the FFI stuff removed.

Paragraphs 3, 4, 5, and 6 in the new version ("The `unsafe` keyword..." to "There is also the...") are a rewording of paragraphs 4, 5, and 6 from the original. Paragraph 3, which is one sentence, describes the two uses of `unsafe`. Paragraphs 4 and 5 provide greater detail on the two uses, and paragraph 6 notes the one special case.

Paragraphs 7, 8, 9, and 10 in the new version ("The standard library has..." to "`Send` is a marker trait") are almost identical to paragraphs 7, 8, 9, and 10 in the original. They give examples of unsafe APIs in the standard and core libraries.

Paragraph 11 in the new version ("Much of the Rust standard library...") is pulled from paragraph 13 in the original. This just explains that the use of unsafe in the standard library isn't a cause for alarm.

Paragraphs 12 and 13 in the new version ("The need for all this..." and "No matter what...") are a rewording of paragraphs 11 and 12 in the original. These introduce probably the biggest point in the piece, the core idea that differentiates Safe Rust and Unsafe Rust: that Safe Rust can't cause undefined behavior.

Paragraphs 14-18 in the new version ("The design of the...") are a reworking of paragraphs 13-21 in the original. These provide greater detail about when to use unsafe, and how unsafe code must be written not to rely on safe code.

Paragraph 19 in the new version ("The decision of whether...") is a reworking of paragraphs 21-23 in the original. This tries to explain that use of `unsafe` is a design choice.

Paragraphs 20 and 21 in the new version are new, and not adapted from the original. Paragraph 20 clarifies how `Send` and `Sync` work, and paragraph 21 provides a small conclusion to the chapter.

Hope that helps!

**Comment by steveklabnik** [comment]
@bors: r+ rollup

thanks a ton!

**Comment by AndrewBrinker** [comment]
:smile:
